### PR TITLE
Add new get_catalog_relations macro, Supporting Changes

### LINKED ISSUE #8521: [CT-3050] [Applied State] Implement Core Changes for Filtered Catalog Queries
### Short description

Building on some of the research done for #4997, we want to improve catalog creation performance by selecting only the needed relation information, instead of information for every relation in a schema.

### Acceptance criteria

- In dbt's base adapter implementation (base/impl.py), the `_get_one_catalog` method has been modified to accept an additional new parameter, `relations_by_schema: Optional[Dict[str, Optional[List[BaseRelation]]]] = None`, mapping schema names to the list of relations in that schema for which catalog information should be retrieved.
- When the new parameter defaults to None, behavior should be the same as before.
- When `_get_one_catalog` invokes the `get_catalog` macro, it should include the new parameter.
- Adapters which do not yet implement the new get_catalog macro parameter should continue to work as before. Therefore we need to evaluate the macro without generating an error and we also need to continue post-filtering the results.
- When the value of the `relations` parameter is `None`, the `_get_one_catalog` method, and `get_catalog` macros should continue to retrieve all relation information for every schema listed in the `schema` parameter, matching historical behavior.

### Impact to Other Teams

The adapter teams will need to implement changes to the get_catalog macro in each adapter implementation, in order to take advantage of the new parameter.

### PR DESCRIPTION
**Pull request by peterallenwebb**
resolves #8521

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: core/dbt/include/global_project/macros/adapters/metadata.sql
Comment (mikealfare) [inline core/dbt/include/global_project/macros/adapters/metadata.sql:1]: I like this approach for opting-in to the new approach. But if we update the macro name in `adapter/base/relation.py`, does it matter? Do we get what we're looking for? I'm not sure how often adapter maintainers actually overwrite that, so overwriting the name in `adapter/base/relation.py` may have the same effect as just updating the contents of the `get_catalog` macro.

Also, if we're breaking this out as a separate macro and making relations_by_schema a `Dict[str, Optional[List[BaseRelation]]]` (emphasis on `BaseRelation` vs `str`), then couldn't we make it just `List[BaseRelation]]`? `BaseRelation` has a `schema` property already, so doing the work of organizing by schema may be redundant.

**Review by mikealfare — COMMENTED** [review]
I read through your draft and have some comments. Most of them are just thoughts that I'm throwing against the wall, so take them with a grain of salt; I'm not strongly for or against anything that I said.

File: core/dbt/adapters/base/impl.py
Comment (peterallenwebb) [reply core/dbt/adapters/base/impl.py:76]: Only that the name of the macro has also changed, so I didn't want the name to be confusing.

File: core/dbt/adapters/base/impl.py
Comment (peterallenwebb) [reply core/dbt/adapters/base/impl.py:1086]: I agree with most of what you've said, and considered some of the same ways out of the problem. There is a complication, however, which is that many adapters override this function. Upon reflection, the approach I took doesn't really fix that in the longer term, so I'll give it some more thought.

**Review by gshank — APPROVED** [review]
My only concern is the lines that show up as untested, which looks like it mainly happens because none of our test cases contain more than 100 relations. I assume that part was hand tested? I wonder if we should create a large test case... but I don't think we need to hold this up for that.

Otherwise looks good.

File: core/dbt/adapters/base/impl.py
Comment (mikealfare) [inline core/dbt/adapters/base/impl.py:1145]: A few observations:

- We're calling `self._get_catalog_relations()` twice, once for the length and again inside of `self._get_catalog_relations_by_info_schema`
- `self._get_catalog_relations_by_info_schema` is doing two things, getting the relations and reformatting the resulting list into a dictionary
- The new method `self._get_catalog_relations_by_info_schema` is only used once, here
- `relations_by_schema` is grouping by the information_schema, not `relation.schema`, which is misleading
- `relations_by_schema` is created to filter a list of relations

Perhaps we could simplify this by reusing the results of `self._get_catalog_relations(manifest)` (let's refer to this as `relations`) and filtering `relations` by `info_schema` while looping through the `info_schema` values?

Some pseudo code to add context (and to organize my own thoughts):

```python
def get_catalog(self, manifest):
    with executor(self.config) as tpe:
        futures = []
        relations = self._get_catalog_relations(manifest)
        if self.CATALOG_BY_RELATION_SUPPORT and len(relations) <= 100:
            info_schemas = {relation.information_schema_only() for relation in relations}
            for info_schema in info_schemas:
                name = ".".join([str(info_schema.database), "information_schema"])
                relations_in_this_info_schema = [
                    r for r in relations if r.information_schema_only() == info_schema
                ],
                fut = tpe.submit_connected(
                    self,
                    name,
                    self._get_one_catalog_by_relations,
                    info_schema,
                    relations_in_this_info_schema,
                    manifest,
                )
                futures.append(fut)
        else:
            # old way
```

File: core/dbt/adapters/base/impl.py
Comment (mikealfare) [inline core/dbt/adapters/base/impl.py:1117]: What does the name of this macro mean? Put another way, how do I know what this should return? It looks like this runs the new macro, but also limits it to objects in the manifest; is that right? This only gets called once, and in the branch of code where we're passing in an explicit list of relations that we got from the manifest. Is the filter still necessary?

I think there's value in wrapping the macro itself in an adapter method. But if we do need to filter it for some reason, then we should do that after this is called.

**Review by mikealfare — COMMENTED** [review]
I left a bunch of comments; let me know if you'd like to discuss any of them. Also, this PR lacks a changelog, do we need one?

File: plugins/postgres/dbt/include/postgres/macros/catalog.sql
Comment (peterallenwebb) [reply plugins/postgres/dbt/include/postgres/macros/catalog.sql:50]: The `schemas` parameter to this macro is indeed a list of strings. The loop below converts it to relations: List[obj] where the obj is acting like a BaseRelation (specifically a "schema relation") for the limited purpose of calling `get_catalog_where_clause`. 

So when this macro is called with schemas = [ "schema_a", "schema_b", "schema_c" ], it will in turn call `get_catalog_where_clause` with relations = [ { "schema": "schema_a" }, { "schema": "schema_b" }, { "schema": "schema_c" } ].

This is just to make it easier to reuse get_catalog_where_clause.

File: core/dbt/adapters/base/relation.py
Comment (peterallenwebb) [reply core/dbt/adapters/base/relation.py:462]: Yes. That's correct. I just fixed it because it was bothering me while I was reverse-engineering and debugging this code.

File: plugins/postgres/dbt/include/postgres/macros/catalog.sql
Comment (mikealfare) [reply plugins/postgres/dbt/include/postgres/macros/catalog.sql:50]: Oh, does jinja treat `relation.schema` the same as `relation.get("schema")` or `relation["schema"]`?

File: core/dbt/adapters/base/impl.py
Comment (peterallenwebb) [reply core/dbt/adapters/base/impl.py:428]: Hey, @mikealfare, following this up, since I didn't incorporate every one of your requested changes. 

My thinking at the time was that the code I had was tested and working, so I didn't want continue tweaking relatively small details after we had already done a lot of collaboration and gone multiple rounds of review. I should have at least left a note to that effect.

However, if you think this or any of the other remaining issues ought to be addressed before 1.7.0, let me know and I am happy to open a follow up issue.

File: core/dbt/adapters/base/impl.py
Comment (peterallenwebb) [reply core/dbt/adapters/base/impl.py:1151]: For this function, it may be worth reviewing the state of the code today if you still have concerns. Gerda refactored it a bit, and made changes along the lines you are suggesting.

**Pull request merged**
